lib/modules/manager/sbt/extract.ts

```diff
@@ -335,11 +385,121 @@ export function extractPackageFile(
+      deps: [],
+      registryUrls,
+      localVars: globalVars,
+      globalVars: {},
+      packageFile,
+    };
+    const res = extractPackageFile(content, acc);
+
+    if (res) {
+      globalVars = { ...globalVars, ...res.localVars };
+      if (res.registryUrls) {
+        registryUrls = [...registryUrls, ...new Set(res.registryUrls)];
+      }
+      if (res.scalaVersion) {
+        scalaVersion = res.scalaVersion;
+      }
+    }
+  }
+
+  return {
+    globalVars,
+    registryUrls,
+    scalaVersion,
+  };
+}
+
+export async function extractAllPackageFiles(
+  _config: ExtractConfig,
+  packageFiles: string[]
+): Promise<PackageFile[] | null> {
+  // Read packages and store in groupPackageFileContent
+  // group package file by its folder
+  const groupPackageFileContent: GroupFilenameContent = {};
+  for (const packageFile of packageFiles) {
+    const content = await readLocalFile(packageFile, 'utf8');
+    if (!content) {
+      logger.trace({ packageFile }, 'packageFile has no content');
+      continue;
+    }
+    const group = upath.dirname(packageFile);
+    groupPackageFileContent[group] ??= [];
+    groupPackageFileContent[group].push({ packageFile, content });
+  }
+
+  // 1. globalVariables from project/ and root package file
+  // 2. registry from all package file
+  // 3. Project's scalaVersion - use in parseDepExpr to add suffix eg. "_2.13"
+  const { globalVars, registryUrls, scalaVersion } = prepareLoadPackageFiles([
+    ...(groupPackageFileContent['project'] ?? []), // in project/ folder
+    ...(groupPackageFileContent['.'] ?? []), // root
+  ]);
+
+  const mapDepsToPackageFile: Record<string, PackageDependency[]> = {};
+  // Start extract all package files
+  for (const packageFileContents of Object.values(groupPackageFileContent)) {
+    // Extract package file by its group
+    // local variable is share within its group
+    for (const { packageFile, content } of packageFileContents) {
+      const res = extractPackageFile(content, {
+        registryUrls,
+        deps: [],
+        packageFile,
+        scalaVersion,
+        localVars: {},
+        globalVars,
+      });
+      if (res) {
+        if (res?.deps) {
+          for (const dep of res.deps) {
+            const variableSourceFile = dep?.editFile ?? packageFile;
+            dep.registryUrls = [...new Set(dep.registryUrls)];
+            mapDepsToPackageFile[variableSourceFile] ??= [];
+            mapDepsToPackageFile[variableSourceFile].push(dep);
+          }
+        }
+      }
+    }
   }
 
-  return { deps, packageFileVersion };
+  // Filter unique package
+  // As we merge all package to single package file
+  // Packages are counted in submodule but it's the same one
+  // by packageName and currentValue
+  const finalPackages = Object.entries(mapDepsToPackageFile).map(
+    ([packageFile, deps]) => ({
+      packageFile,
+      deps: deps.filter(
+        (val, idx, self) =>
+          idx ===
+          self.findIndex(
```

Comment:
If deduplication here happens by `packageName` and `currentValue`, I understand that a package file can include a combination of these two only once. This also means that only the first `editFile` or `fileReplacePosition` would be replaced.

Reply:
https://github.com/renovatebot/renovate/pull/21622/files#diff-a550300890b64a89e53f84f51ad5d5a439e7fb8c75159f8ea4bbf0641e6729a1R471-R477

I did some changes to simplify this